File: python/graph_helpers.py
"""
Graph helper functions for LeetCode-style graph problems.

Graph is represented as an adjacency list using a dict:
    {1: [2, 3], 2: [4, 5], 4: [5]}

BFS uses a queue (FIFO)  -> level-by-level traversal
DFS uses a stack (LIFO)  -> deep traversal down one path first

Time:  O(V + E) - every vertex is popped/processed once,
       every edge is processed once when extending the queue/stack
Space: O(V) - up to all vertices in the queue/stack/traversal list
"""

from collections import deque


class Graph:
    def __init__(self):
        self.graph = {}

    def add_edge(self, from_vertex, to_vertex, bidirectional=False):
        # add edge
        if from_vertex in self.graph:
            self.graph[from_vertex].append(to_vertex)
        else:
            self.graph[from_vertex] = [to_vertex]

        # make sure to_vertex has an entry too (even if empty)
        # so lookups like `vertex in self.graph` behave consistently
        if to_vertex not in self.graph:
            self.graph[to_vertex] = []

        if bidirectional:
            self.graph[to_vertex].append(from_vertex)

    def bfs(self, start_vertex):
        # Check if the start_vertex exists in the graph
        if start_vertex not in self.graph:
            return []  # If not, return an empty list

        queue = deque([start_vertex])  # Initialize a queue with the start_vertex
        visited = {start_vertex}       # track visited to avoid re-adding to queue
        traversal = []                 # Initialize an empty list to store the BFS traversal result

        while queue:  # Continue until the queue is empty
            vertex = queue.popleft()  # Remove and retrieve the first vertex from the queue (FIFO)
            traversal.append(vertex)  # Add the vertex to the traversal result

            for neighbor in self.graph.get(vertex, []):  # Check the vertex's neighbors
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)  # Add unvisited neighbors to the queue

        return traversal  # Return the BFS traversal result

    def dfs(self, start_vertex):
        # Check if the start_vertex exists in the graph
        if start_vertex not in self.graph:
            return []  # If not, return an empty list

        stack = [start_vertex]  # Initialize a stack with the start_vertex
        visited = set()
        traversal = []          # Initialize an empty list to store the DFS traversal result

        while stack:  # Continue until the stack is empty
            vertex = stack.pop()  # Remove and retrieve the last vertex from the stack (LIFO)
            if vertex not in visited:  # Check if the vertex has not been visited
                visited.add(vertex)
                traversal.append(vertex)  # Add the vertex to the traversal result
                # reversed() keeps neighbor visit order consistent with BFS
                stack.extend(reversed(self.graph.get(vertex, [])))

        return traversal  # Return the DFS traversal result

    def bfs_levels(self, start_vertex):
        """Bonus helper: BFS but grouped by level/depth.
        Useful for questions like 'minimum depth' or 'average per level'.
        Returns a list of lists, one per level.
        """
        if start_vertex not in self.graph:
            return []

        queue = deque([start_vertex])
        visited = {start_vertex}
        levels = []

        while queue:
            level_size = len(queue)
            level = []
            for _ in range(level_size):
                vertex = queue.popleft()
                level.append(vertex)
                for neighbor in self.graph.get(vertex, []):
                    if neighbor not in visited:
                        visited.add(neighbor)
                        queue.append(neighbor)
            levels.append(level)

        return levels

    def has_cycle(self):
        """Detect a cycle in a directed graph using DFS + recursion stack.
        Use this version for DIRECTED graphs.
        """
        visited = set()
        in_stack = set()

        def dfs_visit(vertex):
            visited.add(vertex)
            in_stack.add(vertex)
            for neighbor in self.graph.get(vertex, []):
                if neighbor not in visited:
                    if dfs_visit(neighbor):
                        return True
                elif neighbor in in_stack:
                    return True
            in_stack.remove(vertex)
            return False

        for vertex in list(self.graph):
            if vertex not in visited:
                if dfs_visit(vertex):
                    return True
        return False

    def find_largest_node(self, start_vertex):
        # Iterative BFS - traverses entire graph keeping track of the
        # largest node and making comparisons at each node.
        # Starting at a node with no edges to traverse returns None -
        # which is expected (not a problem for non-directed graphs).
        if start_vertex not in self.graph:
            return None

        queue = [start_vertex]
        traversal = []
        largest_node = start_vertex

        while queue:
            vertex = queue.pop(0)

            if vertex not in traversal:
                traversal.append(vertex)

                if vertex > largest_node:
                    largest_node = vertex

                if vertex in self.graph:
                    queue.extend(self.graph[vertex])

        return largest_node

    def find_cycle(self, start_vertex):
        """Iterative DFS - stack to traverse graph and visited set to
        identify a cycle. Use this version for UNDIRECTED graphs.
        The parent check is necessary here so we don't treat the edge
        we just came from as a cycle; omit it for directed graphs
        (use has_cycle instead).
        """
        if start_vertex not in self.graph:
            return False  # No cycle if the start vertex is not in the graph

        stack = [(start_vertex, -1)]  # Stack of (vertex, parent)
        visited = set()  # Set to store visited vertices

        while stack:
            vertex, parent = stack.pop()
            if vertex in visited:
                return True  # A cycle is found if the vertex is already visited

            visited.add(vertex)

            # Get neighbors, or an empty list if the vertex is not explicitly listed
            for neighbor in self.graph.get(vertex, []):
                if neighbor != parent:
                    stack.append((neighbor, vertex))

        return False

    def count_edges(self):
        """Easily count the total number of edges in the graph by
        iterating through each vertex and summing up the lengths of
        their adjacency lists.
        """
        edge_count = 0  # Initialize the edge count to zero
        for vertex in self.graph:  # iterate over each vertex in the graph
            edge_count += len(self.graph[vertex])  # add the number of edges from this vertex
        return edge_count


if __name__ == "__main__":
    g = Graph()
    g.add_edge(1, 2)
    g.add_edge(1, 3)
    g.add_edge(2, 4)
    g.add_edge(2, 5)
    g.add_edge(4, 5)

    print("Graph:", g.graph)
    print("BFS from 1:", g.bfs(1))
    print("DFS from 1:", g.dfs(1))
    print("BFS levels from 1:", g.bfs_levels(1))
    print("Has cycle (directed):", g.has_cycle())
    print("Edge count:", g.count_edges())
    print("Largest node from 1:", g.find_largest_node(1))

    # find_cycle expects an undirected graph (edges added both ways)
    undirected = Graph()
    undirected.add_edge(1, 2, bidirectional=True)
    undirected.add_edge(2, 3, bidirectional=True)
    print("\nUndirected graph:", undirected.graph)
    print("Find cycle (no cycle yet):", undirected.find_cycle(1))

    undirected.add_edge(3, 1, bidirectional=True)  # closes the loop 1-2-3-1
    print("Undirected graph with cycle:", undirected.graph)
    print("Find cycle (cycle present):", undirected.find_cycle(1))

File: src/day1/graphHelpers.ts
import { AdjacencyList } from "@code/types";

/**
 * Graph BFS/DFS helpers for index-based adjacency-list graphs
 * (AdjacencyList = number[][], each index is a vertex, each entry
 * is a list of neighbor indices).
 *
 * Two categories of function here, don't confuse them:
 *  - Pathfinders (dfs, bfs): start -> target, return the path or
 *    null if target is unreachable.
 *  - Traversals (dfsTraversal, bfsTraversal): visit everything
 *    reachable from start, no target, return visit order.
 *
 * All operations below are O(V + E) time, O(V) space unless noted
 * otherwise - V = vertices, E = edges, each visited/processed once.
 */

// ------------------------------------------------------------
// Pathfinders
// ------------------------------------------------------------

/**
 * Depth-first search for a path from `start` to `target`.
 * Returns *a* path, not necessarily the shortest - depends on
 * neighbor order in each vertex's list. Use bfs for shortest path.
 */
export function dfs(
  list: AdjacencyList,
  start: number,
  target: number
): number[] | null {
  if (start < 0 || start >= list.length) return null;

  const visited = new Set<number>();
  const path: number[] = [];

  function explore(node: number): boolean {
    visited.add(node);
    path.push(node);

    if (node === target) return true;

    for (const neighbor of list[node] ?? []) {
      if (!visited.has(neighbor) && explore(neighbor)) {
        return true;
      }
    }

    path.pop(); // backtrack - this node doesn't lead to target
    return false;
  }

  return explore(start) ? path : null;
}

/**
 * Breadth-first search for a path from `start` to `target`.
 * Returns the SHORTEST path (fewest edges), or null if unreachable.
 */
export function bfs(
  list: AdjacencyList,
  start: number,
  target: number
): number[] | null {
  if (start < 0 || start >= list.length) return null;

  const queue: number[] = [start];
  const visited = new Set<number>([start]);
  const parent = new Map<number, number>();

  while (queue.length) {
    const node = queue.shift()!;

    if (node === target) {
      return reconstructPath(parent, start, target);
    }

    for (const neighbor of list[node] ?? []) {
      if (!visited.has(neighbor)) {
        visited.add(neighbor);
        parent.set(neighbor, node);
        queue.push(neighbor);
      }
    }
  }

  return null;
}

function reconstructPath(
  parent: Map<number, number>,
  start: number,
  target: number
): number[] {
  const path: number[] = [target];
  let current = target;

  while (current !== start) {
    current = parent.get(current)!;
    path.push(current);
  }

  return path.reverse();
}

// ------------------------------------------------------------
// Traversals (no target - visit everything reachable from start)
// ------------------------------------------------------------

/** Breadth-first traversal - visit order is level by level. */
export function bfsTraversal(list: AdjacencyList, start: number): number[] {
  if (start < 0 || start >= list.length) return [];

  const queue: number[] = [start];
  const visited = new Set<number>([start]);
  const traversal: number[] = [];

  while (queue.length) {
    const node = queue.shift()!;
    traversal.push(node);

    for (const neighbor of list[node] ?? []) {
      if (!visited.has(neighbor)) {
        visited.add(neighbor);
        queue.push(neighbor);
      }
    }
  }

  return traversal;
}

/**
 * Depth-first traversal - iterative (stack-based), so it can't blow
 * the call stack on a large/degenerate graph. Neighbors are pushed
 * in reverse so the FIRST neighbor in each vertex's list is still
 * explored first, matching what recursive DFS would visit.
 */
export function dfsTraversal(list: AdjacencyList, start: number): number[] {
  if (start < 0 || start >= list.length) return [];

  const stack: number[] = [start];
  const visited = new Set<number>();
  const traversal: number[] = [];

  while (stack.length) {
    const node = stack.pop()!;

    if (visited.has(node)) continue;
    visited.add(node);
    traversal.push(node);

    const neighbors = list[node] ?? [];
    for (let i = neighbors.length - 1; i >= 0; i--) {
      if (!visited.has(neighbors[i])) {
        stack.push(neighbors[i]);
      }
    }
  }

  return traversal;
}

// ------------------------------------------------------------
// Core operations
// ------------------------------------------------------------

/**
 * Find the largest vertex value reachable from `start` (BFS).
 * Returns `start` itself if nothing bigger is reachable.
 */
export function findLargestNode(list: AdjacencyList, start: number): number | null {
  if (start < 0 || start >= list.length) return null;

  const queue: number[] = [start];
  const visited = new Set<number>([start]);
  let largest = start;

  while (queue.length) {
    const node = queue.shift()!;
    if (node > largest) largest = node;

    for (const neighbor of list[node] ?? []) {
      if (!visited.has(neighbor)) {
        visited.add(neighbor);
        queue.push(neighbor);
      }
    }
  }

  return largest;
}

/**
 * Detect a cycle in a DIRECTED graph using DFS + a "currently on
 * the recursion stack" set. A back-edge to a node still on the
 * stack means a cycle; a back-edge to a node that's fully finished
 * (visited but off the stack) is just a re-converging path.
 */
export function hasCycleDirected(list: AdjacencyList): boolean {
  const visited = new Set<number>();
  const onStack = new Set<number>();

  function visit(node: number): boolean {
    visited.add(node);
    onStack.add(node);

    for (const neighbor of list[node] ?? []) {
      if (!visited.has(neighbor)) {
        if (visit(neighbor)) return true;
      } else if (onStack.has(neighbor)) {
        return true;
      }
    }

    onStack.delete(node);
    return false;
  }

  for (let vertex = 0; vertex < list.length; vertex++) {
    if (!visited.has(vertex) && visit(vertex)) return true;
  }

  return false;
}

/**
 * Detect a cycle in an UNDIRECTED graph (edges present both ways
 * in the list). Needs a parent check, unlike the directed version -
 * otherwise walking straight back along the edge you just arrived
 * on would be flagged as a false cycle.
 */
export function hasCycleUndirected(list: AdjacencyList): boolean {
  const visited = new Set<number>();

  function visit(node: number, parent: number): boolean {
    visited.add(node);

    for (const neighbor of list[node] ?? []) {
      if (!visited.has(neighbor)) {
        if (visit(neighbor, node)) return true;
      } else if (neighbor !== parent) {
        return true;
      }
    }

    return false;
  }

  for (let vertex = 0; vertex < list.length; vertex++) {
    if (!visited.has(vertex) && visit(vertex, -1)) return true;
  }

  return false;
}

/**
 * Count total edges by summing adjacency list lengths.
 * For a DIRECTED graph this is the true edge count. For an
 * UNDIRECTED graph (edges stored both ways) this counts each edge
 * twice - divide by 2 if you need the undirected count.
 */
export function countEdges(list: AdjacencyList): number {
  let edgeCount = 0;
  for (const neighbors of list) {
    edgeCount += neighbors.length;
  }
  return edgeCount;
}
